12-offer-encoding.md

```diff
@@ -0,0 +1,840 @@
+    2. data:
+        * [`tu64`:`msat`]
+    1. type: 84 (`invreq_features`)
+    2. data:
+        * [`...*byte`:`features`]
+    1. type: 86 (`invreq_quantity`)
+    2. data:
+        * [`tu64`:`quantity`]
+    1. type: 88 (`invreq_payer_id`)
+    2. data:
+        * [`point`:`key`]
+    1. type: 89 (`invreq_payer_note`)
+    2. data:
+        * [`...*utf8`:`note`]
+    1. type: 240 (`signature`)
+    2. data:
+        * [`bip340sig`:`sig`]
+
+## Requirements for Invoice Requests
+
+The writer:
+  - if it is responding to an offer:
+    - MUST copy all fields from the offer (including unknown fields).
+    - if `offer_chains` is set:
+      - MUST set `invreq_chain` to one of `offer_chains` unless that chain is bitcoin, in which case it MAY omit `invreq_chain`.
+    - otherwise:
+      - if it sets `invreq_chain` it MUST set it to bitcoin.
+    - MUST set `signature`.`sig` as detailed in [Signature Calculation](#signature-calculation) using the `invreq_payer_id`.
+    - if `offer_amount` is not present:
+      - MUST specify `invreq_amount`.
+    - otherwise:
+      - MAY omit `invreq_amount`.
+      - if it sets `invreq_amount`:
+        - MUST specify `invreq_amount`.`msat` as greater or equal to amount expected by `offer_amount` (and, if present, `offer_currency` and `invreq_quantity`).
+    - MUST set `invreq_payer_id` to a transient public key.
+    - MUST remember the secret key corresponding to `invreq_payer_id`.
+    - if `offer_quantity_max` is present:
+      - MUST set `invreq_quantity` to greater than zero.
+      - if `offer_quantity_max` is non-zero:
+        - MUST set `invreq_quantity` less than or equal to `offer_quantity_max`.
+    - otherwise:
+      - MUST NOT set `invreq_quantity`
+  - otherwise (not responding to an offer):
+    - MUST set (or not set) `offer_description`, `offer_absolute_expiry`, `offer_paths` and `offer_issuer` as it would for an offer.
+    - MUST set `invreq_payer_id` as it would set `offer_node_id` for an offer.
+    - MUST NOT include `signature`, `offer_metadata`, `offer_chains`, `offer_amount`, `offer_currency`, `offer_features`, `offer_quantity_max` or `offer_node_id`
+    - if the chain for the invoice is not solely bitcoin:
+      - MUST specify `invreq_chain` the offer is valid for.
+    - MUST set `invreq_amount`.
+  - MUST NOT set any non-signature TLV fields greater or equal to 160.
+  - MUST set `invreq_metadata` to an unpredictable series of bytes.
+  - if it sets `invreq_amount`:
+    - MUST set `msat` in multiples of the minimum lightning-payable unit
+        (e.g. milli-satoshis for bitcoin) for `invreq_chain` (or for bitcoin, if there is no `invreq_chain`).
+  - if it supports bolt12 invoice request features:
+    - MUST set `invreq_features`.`features` to the bitmap of features.
+
+The reader:
+  - MUST fail the request if `invreq_payer_id` or `invreq_metadata` are not present.
+  - MUST fail the request if any non-signature TLV fields greater or equal to 160.
```

Comment:
Should we add a similar constraint on `invoice`?

Reply:
This constraint is because TLVs 160 and up are reserved for invoices so there is no similar constraint to add on invoice.
An invoice request takes an offer and adds fields to it, an invoice takes an invoice request and adds fields to it, but nothing comes after the invoice to add fields to it.

Reply:
Not quite true. 240 to 1000 are reserved for signatures, which an invoice request has, too. If anything above 1000 was also for invoices, then invoices would need two ranges, which is less ideal.

Reply:
> 240 to 1000 are reserved for signatures
There is indeed a line that says that but then only 240 is used and there is not even a hint as to what another signature would be. I believe that it's just here for historical reasons and could (should?) be removed, today there is only one signature field, it's 240. I would be interested to know if anyone is planning on using other signature fields and for what.

Regardless, I don't think it's problematic to allow invoices to use TLVs above 1000, they don't *need* two ranges. And it's one less check.

Reply:
> > 240 to 1000 are reserved for signatures
> 
> There is indeed a line that says that but then only 240 is used and there is not even a hint as to what another signature would be. I believe that it's just here for historical reasons and could (should?) be removed, today there is only one signature field, it's 240. I would be interested to know if anyone is planning on using other signature fields and for what.

I'm not aware of any historical reasons unless they are inside @rustyrussell's brain. Will let him respond. Probably should have a reason in the spec as to why there is though.

> Regardless, I don't think it's problematic to allow invoices to use TLVs above 1000, they don't _need_ two ranges. And it's one less check.

Depends on how you've implemented it. 😛

Reply:
There are many other self-verifying types we can add in future.

For example, if you want to allow a bolt12 invoice to be converted into a bolt11 (given some mapping scheme), you would want a bolt11-sig field.  If you wanted to provide a partial invoice or invoice requests, you would include a field with the missing merkle leaves so you could still check the signature.  Of course, alternate signature schemes are possible too.

You may want to append data, either internally or as an optional extension: this region allows that, without altering the signature.

Reply:
But there is a valid question on future or experimental extensions:
- Can you use 1000+ for them? 
- Should we similarly divide this range?  (Logically, it would be 1000-22511, 22512-44023, 44025-65535).  We could do the same for the other ranges if we really wanted
- Should you echo fields you don't understand?  I think yes, and rely on feature bits to indicate whether you "got it" or not?

Reply:
Yeah, I'd prefer to divide the range and echo. Though note that 1000 is part of the signature range.